okapi-core/src/main/java/org/folio/okapi/util/TokenCache.java

```diff
@@ -0,0 +1,174 @@
+package org.folio.okapi.util;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.OkapiLogger;
+
+public class TokenCache {
+
+  public static final long DEFAULT_TTL = 3 * 60 * 1000L;
+  public static final int DEFAULT_MAX_SIZE = 10000;
+
+  private final Map<String, CacheEntry> cache;
+  private final long ttl;
+
+  private static final Logger logger = OkapiLogger.get(TokenCache.class);
+
+  /**
+   * Constructor using the provided TTL and maxSize.
+   * 
+   * @param ttl cache entry time to live in milliseconds
+   * @param maxSize the maximum number of entries that may be cached at once
+   */
+  private TokenCache(long ttl, int maxSize) {
+    logger.info("Initializing token cache w/ ttl: {}, maxSize: {}", ttl, maxSize);
+    this.ttl = ttl;
+
+    this.cache = Collections.synchronizedMap(new LruCache<>(maxSize));
+  }
+
+  /**
+   * Cache an entry.
+   * 
+   * @param method HTTP method
+   * @param path path pattern
+   * @param userId X-Okapi-User-Id header to cache
+   * @param xokapiPerms X-Okapi-Permissions header to cache
+   * @param keyToken the token to be used in the cache key, from the request which triggered the
+   *        call to mod-authtoken
+   * @param token access token to cache - from the mod-authtoken response
+   */
+  public void put(String tenant, String method, String path, String userId, String xokapiPerms,
+      String keyToken, String token) {
+    long now = System.currentTimeMillis();
+    CacheEntry entry = new CacheEntry(token, userId, xokapiPerms, now + ttl);
+    String key = genKey(method, path, userId, keyToken);
```

Comment:
Since `userId` is part of the key, is there a reason to store it as part of the `CacheEntry` as well?

Reply:
yeah I guess not.  Adam and Julian have both asked questions about the cache keys as well, I guess this needs more thought

Reply:
I think getting rid of user will be best. That means whatever content mod-authtoken puts there in the future is "safe".

Reply:
Removed userId from the cache key